starboard/nplb/player_write_sample_test.cc

```diff
@@ -102,16 +103,176 @@ TEST_P(SbPlayerWriteSampleTest, WriteMultipleBatches) {
+    return;
+  }
+
+  // TODO: we simply set audio write duration to 0.5 second. Ideally, we should
+  // set the audio write duration to 10 seconds if audio connectors are remote.
+  player_fixture.SetAudioWriteDuration(kSbTimeSecond / 2);
+
+  GroupedSamples samples;
+  if (player_fixture.HasAudio()) {
+    samples.AddAudioSamples(
+        0, player_fixture.ConvertDurationToAudioBufferCount(kSbTimeSecond));
+    samples.AddAudioEOS();
   }
   if (player_fixture.HasVideo()) {
-    samples.AddVideoSamplesWithEOS(0, samples_to_write);
+    samples.AddVideoSamples(
+        0, player_fixture.ConvertDurationToVideoBufferCount(kSbTimeSecond));
+    samples.AddVideoEOS();
   }
 
   ASSERT_NO_FATAL_FAILURE(player_fixture.Write(samples));
   ASSERT_NO_FATAL_FAILURE(player_fixture.WaitForPlayerEndOfStream());
 }
 
+TEST_P(SbPlayerWriteSampleTest, PartialAudio) {
+  if (!IsPartialAudioSupported()) {
+    // TODO: Use GTEST_SKIP when we have a newer version of gtest.
+    SB_LOG(INFO)
+        << "The platform doesn't support partial audio. Skip the tests.";
+    return;
+  }
+
+  SbPlayerTestFixture player_fixture(GetParam());
+  if (HasFatalFailure()) {
+    return;
+  }
+  if (!player_fixture.HasAudio()) {
+    // TODO: Use GTEST_SKIP when we have a newer version of gtest.
+    SB_LOG(INFO) << "Skip PartialAudio test for audioless content.";
+    return;
+  }
+
+  const SbTime kDurationToPlay = kSbTimeSecond;
+  const float kSegmentSize = 0.1f;
+
+  GroupedSamples samples;
+  if (player_fixture.HasVideo()) {
+    samples.AddVideoSamples(
+        0, player_fixture.ConvertDurationToVideoBufferCount(kDurationToPlay));
+    samples.AddVideoEOS();
+  }
+
+  int total_buffers_to_write =
+      player_fixture.ConvertDurationToAudioBufferCount(kDurationToPlay);
+  for (int i = 0; i < total_buffers_to_write; i++) {
+    SbTime current_timestamp = player_fixture.GetAudioSampleTimestamp(i);
+    SbTime next_timestamp = player_fixture.GetAudioSampleTimestamp(i + 1);
+    SbTime buffer_duration = next_timestamp - current_timestamp;
+    SbTime segment_duration = buffer_duration * kSegmentSize;
+    SbTime written_duration = 0;
+    while (written_duration < buffer_duration) {
+      samples.AddAudioSamples(
+          i, 1, written_duration, written_duration,
+          std::max<int64_t>(
+              0, buffer_duration - written_duration - segment_duration));
+      written_duration += segment_duration;
+    }
+  }
+  samples.AddAudioEOS();
+
+  ASSERT_NO_FATAL_FAILURE(player_fixture.Write(samples));
+  ASSERT_NO_FATAL_FAILURE(player_fixture.WaitForPlayerPresenting());
+
+  SbTime start_system_time = SbTimeGetMonotonicNow();
+  SbTime start_media_time = player_fixture.GetCurrentMediaTime();
+
+  ASSERT_NO_FATAL_FAILURE(player_fixture.WaitForPlayerEndOfStream());
+
+  SbTime end_system_time = SbTimeGetMonotonicNow();
+  SbTime end_media_time = player_fixture.GetCurrentMediaTime();
+
+  const SbTime kDurationDifferenceAllowance = kSbTimeSecond;
```

Comment:
It's great to have a large epsilon to avoid flakiness.  Let's check the duration on a powerful platform (like desktop Linux) to ensure that the actually difference is small (which also helps verifying that the test is implemented correctly).
Also consider logging the values and difference before the ASSERT(), like
"The expected time should be ..., and is ..., with difference ....", please rephrase.